v8.0 - key error (on Windows)

Running gcovr with the following parameters

```bash
python -m pip install gcovr
python -m gcovr -r .. \
  --exclude-noncode-lines \
  --exclude-throw-branches \
  --exclude-unreachable-branches \
  --exclude '.*_deps/.*' \
  --exclude '.*tests/.*' \
  --exclude '.*third-party/.*' \
  --xml-pretty \
  -o coverage.xml
```

pip install logs:
```
Collecting gcovr
  Downloading gcovr-8.0-py2.py3-none-any.whl.metadata (10 kB)
Collecting jinja2 (from gcovr)
  Downloading jinja2-3.1.4-py3-none-any.whl.metadata (2.6 kB)
Collecting lxml (from gcovr)
  Downloading lxml-5.3.0-cp311-cp311-win_amd64.whl.metadata (3.9 kB)
Collecting colorlog (from gcovr)
  Downloading colorlog-6.8.2-py3-none-any.whl.metadata (10 kB)
Collecting pygments>=2.13.0 (from gcovr)
  Downloading pygments-2.18.0-py3-none-any.whl.metadata (2.5 kB)
Collecting colorama (from colorlog->gcovr)
  Downloading colorama-0.4.6-py2.py3-none-any.whl.metadata (17 kB)
Collecting MarkupSafe>=2.0 (from jinja2->gcovr)
  Downloading MarkupSafe-3.0.1-cp311-cp311-win_amd64.whl.metadata (4.1 kB)
Downloading gcovr-8.0-py2.py3-none-any.whl (213 kB)
Downloading pygments-2.18.0-py3-none-any.whl (1.2 MB)
   ---------------------------------------- 1.2/1.2 MB 15.0 MB/s eta 0:00:00
Downloading colorlog-6.8.2-py3-none-any.whl (11 kB)
Downloading jinja2-3.1.4-py3-none-any.whl (133 kB)
Downloading lxml-5.3.0-cp311-cp311-win_amd64.whl (3.8 MB)
   ---------------------------------------- 3.8/3.8 MB 45.5 MB/s eta 0:00:00
Downloading MarkupSafe-3.0.1-cp311-cp311-win_amd64.whl (15 kB)
Downloading colorama-0.4.6-py2.py3-none-any.whl (25 kB)
Installing collected packages: pygments, MarkupSafe, lxml, colorama, jinja2, colorlog, gcovr
  WARNING: The script pygmentize.exe is installed in 'C:\hostedtoolcache\windows\Python\3.11.9\x64\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The script gcovr.exe is installed in 'C:\hostedtoolcache\windows\Python\3.11.9\x64\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed MarkupSafe-3.0.1 colorama-0.4.6 colorlog-6.8.2 gcovr-8.0 jinja2-3.1.4 lxml-5.3.0 pygments-2.18.0
```

I get the following error (https://github.com/LizardByte/Sunshine/actions/runs/11246344220/job/31268055119?pr=3276#step:11:52):
```
(INFO) Reading coverage data...

Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 81, in worker
    work(*args, **kwargs)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 566, in process_datafile
    done = run_gcov_and_process_files(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 922, in run_gcov_and_process_files
    process_gcov_json_data(gcov_filename, covdata, options)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 294, in process_gcov_json_data
    apply_all_exclusions(file_cov, lines=encoded_source_lines, options=options)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\exclusions\__init__.py", line 109, in apply_all_exclusions
    remove_internal_functions(filecov)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\exclusions\__init__.py", line 136, in remove_internal_functions
    filecov.functions.pop(function.demangled_name)
KeyError: '__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > config::skip_list<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >)'
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 81, in worker
    work(*args, **kwargs)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 566, in process_datafile
    done = run_gcov_and_process_files(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 922, in run_gcov_and_process_files
    process_gcov_json_data(gcov_filename, covdata, options)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 294, in process_gcov_json_data
    apply_all_exclusions(file_cov, lines=encoded_source_lines, options=options)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\exclusions\__init__.py", line 109, in apply_all_exclusions
    remove_internal_functions(filecov)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\exclusions\__init__.py", line 136, in remove_internal_functions
    filecov.functions.pop(function.demangled_name)
KeyError: '__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > config::skip_list<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >)'
(ERROR) Error occurred while reading reports:
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\__main__.py", line 358, in main
    covdata: CovData = gcovr_formats.read_reports(options)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\__init__.py", line 85, in read_reports
    covdata = GcovHandler(options).read_report()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\__init__.py", line 215, in read_report
    return read_report(self.options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 96, in read_report
    with Workers(
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 177, in __exit__
    self.wait()
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 168, in wait
    raise self.exceptions[0][1]
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 103, in read_report
    contexts = pool.wait()
               ^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 168, in wait
    raise self.exceptions[0][1]
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 81, in worker
    work(*args, **kwargs)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 566, in process_datafile
    done = run_gcov_and_process_files(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 922, in run_gcov_and_process_files
    process_gcov_json_data(gcov_filename, covdata, options)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 294, in process_gcov_json_data
    apply_all_exclusions(file_cov, lines=encoded_source_lines, options=options)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\exclusions\__init__.py", line 109, in apply_all_exclusions
    remove_internal_functions(filecov)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\exclusions\__init__.py", line 136, in remove_internal_functions
    filecov.functions.pop(function.demangled_name)
KeyError: '__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > config::skip_list<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >)'

'__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > config::skip_list<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >)'
Error: (ERROR) Error occurred while reading reports:
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\__main__.py", line 358, in main
    covdata: CovData = gcovr_formats.read_reports(options)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\__init__.py", line 85, in read_reports
    covdata = GcovHandler(options).read_report()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\__init__.py", line 215, in read_report
    return read_report(self.options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 96, in read_report
    with Workers(
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 177, in __exit__
    self.wait()
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 168, in wait
    raise self.exceptions[0][1]
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 103, in read_report
    contexts = pool.wait()
               ^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 168, in wait
    raise self.exceptions[0][1]
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\workers.py", line 81, in worker
    work(*args, **kwargs)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 566, in process_datafile
    done = run_gcov_and_process_files(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 922, in run_gcov_and_process_files
    process_gcov_json_data(gcov_filename, covdata, options)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\formats\gcov\read.py", line 294, in process_gcov_json_data
    apply_all_exclusions(file_cov, lines=encoded_source_lines, options=options)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\exclusions\__init__.py", line 109, in apply_all_exclusions
    remove_internal_functions(filecov)
  File "C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\site-packages\gcovr\exclusions\__init__.py", line 136, in remove_internal_functions
    filecov.functions.pop(function.demangled_name)
KeyError: '__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > config::skip_list<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >)'

'__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > config::skip_list<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >)'
Error: Process completed with exit code 64.
```

I'm unsure of what changed. Didn't see anything obvious in the release notes.

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -14,15 +14,18 @@ New features and notable changes:
 
 Bug fixes and small improvements:
 
+- Fix exclusion of internal functions. (:issue:`987`)
+
 Documentation:
 
 Internal changes:
-
 
 8.0 (07 October 2024)
 ---------------------
 
 Known bugs:
+
+- Exclusion of internal function not working. (:issue:`984`)
 
 Breaking changes:
 

--- a/gcovr/coverage.py
+++ b/gcovr/coverage.py
@@ -512,7 +512,7 @@ class FileCoverage:
                 "Data for filtering is missing. Need supported GCOV JSON format to get the information."
             )
         filecov = FileCoverage(self.filename)
-        self.functions[functioncov.name] = functioncov
+        filecov.functions[functioncov.name] = functioncov
 
         filecov.lines = {
             line: value

--- a/gcovr/exclusions/__init__.py
+++ b/gcovr/exclusions/__init__.py
@@ -123,8 +123,9 @@ def remove_calls(filecov: FileCoverage):
 def remove_internal_functions(filecov: FileCoverage):
     """Remove compiler-generated functions, e.g. for static initialization."""
 
-    # iterate over shallow copy
-    for function in list(filecov.functions.values()):
+    # Get all the keys first because we want to remove some of them which will else result in an error.
+    for key in list(filecov.functions.keys()):
+        function = filecov.functions[key]
         if _function_can_be_excluded(function.demangled_name):
             LOGGER.debug(
                 "Ignoring symbol %s in line %s in file %s",
@@ -133,7 +134,7 @@ def remove_internal_functions(filecov: FileCoverage):
                 filecov.filename,
             )
 
-            filecov.functions.pop(function.demangled_name)
+            filecov.functions.pop(key)
 
 
 def _function_can_be_excluded(name: str) -> bool:

--- a/gcovr/exclusions/utils.py
+++ b/gcovr/exclusions/utils.py
@@ -149,11 +149,11 @@ def apply_exclusion_ranges(
         elif branch_is_excluded(linecov.lineno):
             linecov.branches = {}
 
-    for functioncov in filecov.functions.values():
-        for lineno in functioncov.excluded.keys():
+    for function in filecov.functions.values():
+        for lineno in function.excluded.keys():
             if line_is_excluded(lineno):
-                functioncov.count[lineno] = 0
-                functioncov.excluded[lineno] = True
+                function.count[lineno] = 0
+                function.excluded[lineno] = True
 
 
 def _make_is_in_any_range_inclusive(

--- a/noxfile.py
+++ b/noxfile.py
@@ -269,6 +269,8 @@ def doc(session: nox.Session) -> None:
         line, _ = re.subn(r"``", r"`", line)
         line, _ = re.subn(r":(?:option|ref):", r"", line)
         line, _ = re.subn(r":issue:`(\d+)`", r"#\1", line)
+        if line.strip().startswith("- ") and out_lines[-1].rstrip() == "":
+            out_lines.pop()
         out_lines.append(line)
     else:
         raise RuntimeError(f"End of release changes not found in {changelog_rst}.")

